### PR TITLE
w8

### PR DESCRIPTION
**Pull request by stevemao**

### CODE REVIEW

File: cis194/week8/stevemao/Party.hs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+moreFun gl1 gl2
+  | compare gl1 gl2 == GT = gl1
+  | otherwise = gl2
+
+treeFold :: (a -> [b] -> b) -> Tree a -> b
+treeFold f (Node a forest) = f a . fmap (treeFold f) $ forest
+
+nextLevel :: Employee -> [(GuestList, GuestList)]
+  -> (GuestList, GuestList)
+nextLevel boss xs = foldr (\(gl1, gl2) (gl1', gl2') -> (glCons boss gl1 <> gl1', gl2 <> gl2')) (mempty, mempty) xs
```
Comment (stevemao) [inline cis194/week8/stevemao/Party.hs]: This logic is wrong... will fix it later.

File: cis194/week8/stevemao/Party.hs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+instance Semigroup GuestList where
+  GL el1 f1 <> GL el2 f2 = GL (el1 <> el2) (f1 + f2)
+
+instance Monoid GuestList where
+  mempty = GL [] 0
+
+moreFun :: GuestList -> GuestList -> GuestList
+moreFun gl1 gl2
+  | compare gl1 gl2 == GT = gl1
+  | otherwise = gl2
```
Comment (shineli1984) [inline cis194/week8/stevemao/Party.hs]: We need something like `Ord a => a -> a -> a`. Do we have a function that we can use under `Ord` typeclass?

File: cis194/week8/stevemao/Party.hs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+treeFold :: (a -> [b] -> b) -> Tree a -> b
+treeFold f (Node a forest) = f a . fmap (treeFold f) $ forest
+
+nextLevel :: Employee -> [(GuestList, GuestList)]
+  -> (GuestList, GuestList)
```
Comment (shineli1984) [inline cis194/week8/stevemao/Party.hs]: We usually align type like this:
```Haskell
nextLevel :: Employee 
          -> [ ( GuestList, GuestList ) ]
          -> ( GuestList, GuestList )
```

File: cis194/week8/stevemao/Party.hs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+nextLevel boss xs = foldr (\(gl1, gl2) (gl1', gl2') -> (glCons boss gl1 <> gl2', gl2 <> gl1')) (mempty, mempty) xs
+
+moreFunTuple :: (GuestList, GuestList) -> GuestList
+moreFunTuple = uncurry moreFun
```
Comment (shineli1984) [inline cis194/week8/stevemao/Party.hs]: If we only use uncurried `moreFunTuple` once, we don't need to bind it to a name. If we have to, we can just use `moreFunTuple'`

File: cis194/week8/stevemao/Party.hs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+moreFun gl1 gl2
+  | compare gl1 gl2 == GT = gl1
+  | otherwise = gl2
+
+treeFold :: (a -> [b] -> b) -> Tree a -> b
+treeFold f (Node a forest) = f a . fmap (treeFold f) $ forest
+
+nextLevel :: Employee -> [(GuestList, GuestList)]
+  -> (GuestList, GuestList)
+nextLevel boss xs = foldr (\(gl1, gl2) (gl1', gl2') -> (glCons boss gl1 <> gl2', gl2 <> gl1')) (mempty, mempty) xs
```
Comment (shineli1984) [inline cis194/week8/stevemao/Party.hs]: Have we tested this logic?

**Pull request merged**